### DISCUSSION TITLE
Proxying to remote host from closed localhost

### DISCUSSION THREAD

**Discussion by DomenPigeon** [dotnet/yarp#2363] [Q&A] [answered]
Is something like this possible to do with YARP? Where Node1 is visible to the outside internet and Node2 is not and is closed down?
 
Node1: 
BlazorApp (with YARP): http://blazor-node1.com

Node2:
MyApp: localhost:1234
YARP Proxy: localhost:1234 -> http://blazor-node1.com/app-on-node2

**Comment by Tratcher** [comment, marked as answer]
Yes, that's fine so long as node 2 can make outbound calls.

**Reply by DomenPigeon** [reply]
@Tratcher Thank you for the answer. Just have one more question:
- Is it also possible to do this dinamically on Node1? So if there are multiple closed down nodes which come and go?

Could you please just point me to documentation or a sample project how to configure this? As far as I have read the docs I imagine that Node2 has direct forwarding configured, but I am not sure how the requests are then fetched on node1.

**Reply by Tratcher** [reply]
What is node 1 using YARP for? If Node 2 is forwarding reqeuests to node 1, node 1 doesn't need YARP to receive them, they come in as normal HTTP requests.

**Reply by DomenPigeon** [reply]
Ok maybe I don' understand something or I have explained something in the wrong way, let me explain the whole problem I am trying to solve:

Node1 is hosting a BlazorApp, has a static ip and is visible to the outside world, let's say on http://blazor-node1.com
Node1 then has a SignalR hub (two way comunication), which accepts other nodes to connect to it, for this no other node needs to have an outside address and can simple connect and disconnect as it pleases.

On Node2 two apps are running: 
- an Asp.net core app monitoring Node2
- some other app on localhost:1234

Asp.net core app on Node2 connects to Node1 via hub and says that it has localhost:1234 app which it wants to redirect to node1.

Node1 then makes http://blazor-node1.com/app-on-node2 accessible, but all the trafic goes to node2 localhost:1234, but here again node2 doesn't have a static address which it can be redirected to.

I imagined that I could somehow route the trafic from Node1 to Node2 via the singnalR hub connection, but then I remembered the YARP project and thought if this is possible to do with it.

**Reply by Tratcher** [reply]
Can node 1 make requests to the internal network (node 2)? If so then you can use direct forwarding or programmatic config for YARP in node 1 to forward requests to node 2. You wouldn't need YARP on node 2 to receive the requests, they'd be normal incoming HTTP requests.

If node 1 can't directly contact node 2 then you need something like a reverse HTTP tunnel opened from node 2 to node 1. It's possible to do with YARP, but it's kind of wacky. I don't know if we have an example around.

**Reply by DomenPigeon** [reply]
Thanks for the reply and sory I didn't manage to clearify what I mean sooner. It's the second case yes, node1 isn't on the same internal network as node2.

As you mentioned http tunnel, I found a feature request which is basically what I had in mind: https://github.com/microsoft/reverse-proxy/issues/1618
And there is even a demo: https://github.com/davidfowl/YarpTunnelDemo

Thanks again for your answers, as I see that this is not (yet) officially suported will check out the demo and try to do it manually.
